Minix/1.7.5/MANUALS/CAT4/HD.4


HD(4)                     Minix Programmer's Manual                      HD(4)


NAME
     hd - winchester hard disk

DESCRIPTION
     The hd* family of devices refer to the Winchester hard disk  drivers  for
     the  IBM XT, AT and PS/2 machines, but may also refer to the generic (and
     slower) BIOS based hard disk driver.  These disks are arrays of 512  byte
     sectors,  although  Minix  always works with two sectors at a time due to
     its 1024 byte block size.  You can read or  write  any  number  of  bytes
     however,  Minix  takes  care  of  cutting  and  pasting incomplete blocks
     together.

     The devices may be divided into three classes:

          The devices with a minor device number that is a multiple of 5, i.e.
          hd0  or  hd5,  refer  to the whole hard disk 0 and 1.  Through these
          devices one has access to any block on the hard disk.  Most  notably
          the  partition  table,  that can be found in the first sector of the
          disk.

          The devices with a minor device number that is not a multiple of  5,
          i.e.   hd1,  hd2,  ..., hd6, ..., refer to primary partitions of the
          lower numbered whole  hard  disk  device.   These  devices  normally
          contain  MS-DOS or Minix file systems.  /dev/hd1 is often the MS-DOS
          C: drive.

          Minor devices from 128 up may refer to  Minix  subpartitions  within
          primary  partitions  if  a  subpartition  table has been placed in a
          Minix primary partition.  The subpartitions of hd3 for instance, are
          named  hd3a  through  hd3d.   Their  minor  device  numbers  may  be
          calculated as 128 + 16*drive + 4*partition + subpartition,  counting
          the partitions from zero.

     If a primary partition is an extended partition then up to  four  logical
     partitions  can  be accessed as subpartitions of that extended partition.
     This allows one  to  access  foreign  file  systems  of  other  operating
     systems,   Minix   file  systems  are  not  normally  placed  in  logical
     partitions.

PARTITIONING
     The first sector of a drive (or partition for  subpartitioning)  contains
     the  partition  table  at byte offset 446.  This is what each of the four
     entries looks like as defined in <ibm/partition.h>:

     /* Description of entry in the partition table.  */

     struct part_entry {
       unsigned char bootind;       /* boot indicator 0/ACTIVE_FLAG      */
       unsigned char start_head;    /* head value for first sector       */
       unsigned char start_sec;     /* sector value + high 2 cyl bits    */
       unsigned char start_cyl;     /* low 8 cylinder bits               */
       unsigned char sysind;        /* system indicator                  */
       unsigned char last_head;     /* h/s/c for the last sector         */
       unsigned char last_sec;
       unsigned char last_cyl;
       unsigned long lowsec;        /* logical first sector              */


                                                                             1



HD(4)                     Minix Programmer's Manual                      HD(4)


       unsigned long size;          /* size of partition in sectors      */
     };

     #define ACTIVE_FLAG     0x80   /* value for active in bootind field */
     #define NR_PARTITIONS   4      /* number of entries in table        */
     #define PART_TABLE_OFF  0x1BE  /* offset of table in boot sector    */

     /* Partition types (sysind). */
     #define MINIX_PART      0x81   /* Minix partition type */
     #define NO_PART         0x00   /* unused entry */
     #define OLD_MINIX_PART  0x80   /* created before 1.4b, obsolete */
     #define EXT_PART        0x05   /* extended partition */

     The cylinder numbers are encoded in a very strange way, bits 8 and 9  are
     in the high two bits of the sector number.  The sector numbers count from
     1, not 0!  More useful are the  lowsec  and  size  fields  however,  they
     simply  give  the  location of the partition as an absolute sector offset
     and length within the drive.

     The partition table entry defined above is specific to  IBM  type  disks.
     The  device  drivers  use  another  partition  entry  structure  to  pass
     information on a partition.  This is what <minix/partition.h> looks like:

     struct partition {
       u64_t base;              /* byte offset to the partition start */
       u64_t size;              /* number of bytes in the partition */
       unsigned cylinders;      /* disk geometry for partitioning */
       unsigned heads;
       unsigned sectors;
     };

     The base and size fields are the byte offset and length of  a  partition.
     (These  are 64 bit numbers under Minix-vmd, but only 32 bit numbers under
     standard Minix.)  The geometry of the disk is also given for the  benefit
     of  partition  table  editors.   This information can be obtained from an
     open disk device with the call:

          ioctl(fd, DIOCGETP, &entry);

     One can change the placement of the device to the lowsec and size  fields
     of  entry  by  using  the  DIOCSETP call instead.  Only the base and size
     fields are used for DIOCSETP.

     The partition tables when read from disk by the driver  are  checked  and
     truncated to fit within the primary partition or drive.  The first sector
     should be left free for the partition table.

     The partition tables are read when the in-use count  (opens  and  mounts)
     changes  from  0 to 1.  So an idle disk is automatically repartitioned on
     the next access.  This  means  that  repartitioning  programs  only  have
     effect  if  a  disk  stays in use, unless they reload a changed partition
     table.






                                                                             2



HD(4)                     Minix Programmer's Manual                      HD(4)


FILES
     /dev/hd[0-9], /dev/hd[1-46-9][a-d]

SEE ALSO
     ioctl(2), int64(3), part(8), repartition(8).

BUGS
     The subpartitioning is incompatible with the MS-DOS  method  of  extended
     partitions.   The  latter  does  not  map well to the sparse minor device
     number space.

     The primary partition  table  is  sorted  by  lowsec  like  MS-DOS  does,
     subpartition  tables  are  not.   Just  think about what happens when you
     delete a partition in the MS-DOS scheme.

     Don't move a partition that is mounted or kept open by some process.  The
     file system may write cached blocks to the new location.

     The BIOS driver is not slow at all on a buffered disk.

     Some IDE disks send an interrupt when they spin down under hardware power
     management.   The  driver acknowledges the interrupt as it is supposed to
     do by reading the status register.  The disk then spins up again...   You
     have to disable the spin down in the computer setup to fix the problem.

AUTHOR
     Kees J. Bot (kjb@cs.vu.nl)































                                                                             3



kernel/0.1x/linux-0.11/kernel/blk_drv/hd.c
/*
 *  linux/kernel/hd.c
 *
 *  (C) 1991  Linus Torvalds
 */

/*
 * This is the low-level hd interrupt support. It traverses the
 * request-list, using interrupts to jump between functions. As
 * all the functions are called within interrupts, we may not
 * sleep. Special care is recommended.
 * 
 *  modified by Drew Eckhardt to check nr of hd's from the CMOS.
 */

#include <linux/config.h>
#include <linux/sched.h>
#include <linux/fs.h>
#include <linux/kernel.h>
#include <linux/hdreg.h>
#include <asm/system.h>
#include <asm/io.h>
#include <asm/segment.h>

#define MAJOR_NR 3
#include "blk.h"

#define CMOS_READ(addr) ({ \
outb_p(0x80|addr,0x70); \
inb_p(0x71); \
})

/* Max read/write errors/sector */
#define MAX_ERRORS	7
#define MAX_HD		2

static void recal_intr(void);

static int recalibrate = 1;
static int reset = 1;

/*
 *  This struct defines the HD's and their types.
 */
struct hd_i_struct {
	int head,sect,cyl,wpcom,lzone,ctl;
	};
#ifdef HD_TYPE
struct hd_i_struct hd_info[] = { HD_TYPE };
#define NR_HD ((sizeof (hd_info))/(sizeof (struct hd_i_struct)))
#else
struct hd_i_struct hd_info[] = { {0,0,0,0,0,0},{0,0,0,0,0,0} };
static int NR_HD = 0;
#endif

static struct hd_struct {
	long start_sect;
	long nr_sects;
} hd[5*MAX_HD]={{0,0},};

#define port_read(port,buf,nr) \
__asm__("cld;rep;insw"::"d" (port),"D" (buf),"c" (nr):"cx","di")

#define port_write(port,buf,nr) \
__asm__("cld;rep;outsw"::"d" (port),"S" (buf),"c" (nr):"cx","si")

extern void hd_interrupt(void);
extern void rd_load(void);

/* This may be used only once, enforced by 'static int callable' */
int sys_setup(void * BIOS)
{
	static int callable = 1;
	int i,drive;
	unsigned char cmos_disks;
	struct partition *p;
	struct buffer_head * bh;

	if (!callable)
		return -1;
	callable = 0;
#ifndef HD_TYPE
	for (drive=0 ; drive<2 ; drive++) {
		hd_info[drive].cyl = *(unsigned short *) BIOS;
		hd_info[drive].head = *(unsigned char *) (2+BIOS);
		hd_info[drive].wpcom = *(unsigned short *) (5+BIOS);
		hd_info[drive].ctl = *(unsigned char *) (8+BIOS);
		hd_info[drive].lzone = *(unsigned short *) (12+BIOS);
		hd_info[drive].sect = *(unsigned char *) (14+BIOS);
		BIOS += 16;
	}
	if (hd_info[1].cyl)
		NR_HD=2;
	else
		NR_HD=1;
#endif
	for (i=0 ; i<NR_HD ; i++) {
		hd[i*5].start_sect = 0;
		hd[i*5].nr_sects = hd_info[i].head*
				hd_info[i].sect*hd_info[i].cyl;
	}

	/*
		We querry CMOS about hard disks : it could be that 
		we have a SCSI/ESDI/etc controller that is BIOS
		compatable with ST-506, and thus showing up in our
		BIOS table, but not register compatable, and therefore
		not present in CMOS.

		Furthurmore, we will assume that our ST-506 drives
		<if any> are the primary drives in the system, and 
		the ones reflected as drive 1 or 2.

		The first drive is stored in the high nibble of CMOS
		byte 0x12, the second in the low nibble.  This will be
		either a 4 bit drive type or 0xf indicating use byte 0x19 
		for an 8 bit type, drive 1, 0x1a for drive 2 in CMOS.

		Needless to say, a non-zero value means we have 
		an AT controller hard disk for that drive.

		
	*/

	if ((cmos_disks = CMOS_READ(0x12)) & 0xf0)
		if (cmos_disks & 0x0f)
			NR_HD = 2;
		else
			NR_HD = 1;
	else
		NR_HD = 0;
	for (i = NR_HD ; i < 2 ; i++) {
		hd[i*5].start_sect = 0;
		hd[i*5].nr_sects = 0;
	}
	for (drive=0 ; drive<NR_HD ; drive++) {
		if (!(bh = bread(0x300 + drive*5,0))) {
			printk("Unable to read partition table of drive %d\n\r",
				drive);
			panic("");
		}
		if (bh->b_data[510] != 0x55 || (unsigned char)
		    bh->b_data[511] != 0xAA) {
			printk("Bad partition table on drive %d\n\r",drive);
			panic("");
		}
		p = 0x1BE + (void *)bh->b_data;
		for (i=1;i<5;i++,p++) {
			hd[i+5*drive].start_sect = p->start_sect;
			hd[i+5*drive].nr_sects = p->nr_sects;
		}
		brelse(bh);
	}
	if (NR_HD)
		printk("Partition table%s ok.\n\r",(NR_HD>1)?"s":"");
	rd_load();
	mount_root();
	return (0);
}

static int controller_ready(void)
{
	int retries=10000;

	while (--retries && (inb_p(HD_STATUS)&0xc0)!=0x40);
	return (retries);
}

static int win_result(void)
{
	int i=inb_p(HD_STATUS);

	if ((i & (BUSY_STAT | READY_STAT | WRERR_STAT | SEEK_STAT | ERR_STAT))
		== (READY_STAT | SEEK_STAT))
		return(0); /* ok */
	if (i&1) i=inb(HD_ERROR);
	return (1);
}

static void hd_out(unsigned int drive,unsigned int nsect,unsigned int sect,
		unsigned int head,unsigned int cyl,unsigned int cmd,
		void (*intr_addr)(void))
{
	register int port asm("dx");

	if (drive>1 || head>15)
		panic("Trying to write bad sector");
	if (!controller_ready())
		panic("HD controller not ready");
	do_hd = intr_addr;
	outb_p(hd_info[drive].ctl,HD_CMD);
	port=HD_DATA;
	outb_p(hd_info[drive].wpcom>>2,++port);
	outb_p(nsect,++port);
	outb_p(sect,++port);
	outb_p(cyl,++port);
	outb_p(cyl>>8,++port);
	outb_p(0xA0|(drive<<4)|head,++port);
	outb(cmd,++port);
}

static int drive_busy(void)
{
	unsigned int i;

	for (i = 0; i < 10000; i++)
		if (READY_STAT == (inb_p(HD_STATUS) & (BUSY_STAT|READY_STAT)))
			break;
	i = inb(HD_STATUS);
	i &= BUSY_STAT | READY_STAT | SEEK_STAT;
	if (i == READY_STAT | SEEK_STAT)
		return(0);
	printk("HD controller times out\n\r");
	return(1);
}

static void reset_controller(void)
{
	int	i;

	outb(4,HD_CMD);
	for(i = 0; i < 100; i++) nop();
	outb(hd_info[0].ctl & 0x0f ,HD_CMD);
	if (drive_busy())
		printk("HD-controller still busy\n\r");
	if ((i = inb(HD_ERROR)) != 1)
		printk("HD-controller reset failed: %02x\n\r",i);
}

static void reset_hd(int nr)
{
	reset_controller();
	hd_out(nr,hd_info[nr].sect,hd_info[nr].sect,hd_info[nr].head-1,
		hd_info[nr].cyl,WIN_SPECIFY,&recal_intr);
}

void unexpected_hd_interrupt(void)
{
	printk("Unexpected HD interrupt\n\r");
}

static void bad_rw_intr(void)
{
	if (++CURRENT->errors >= MAX_ERRORS)
		end_request(0);
	if (CURRENT->errors > MAX_ERRORS/2)
		reset = 1;
}

static void read_intr(void)
{
	if (win_result()) {
		bad_rw_intr();
		do_hd_request();
		return;
	}
	port_read(HD_DATA,CURRENT->buffer,256);
	CURRENT->errors = 0;
	CURRENT->buffer += 512;
	CURRENT->sector++;
	if (--CURRENT->nr_sectors) {
		do_hd = &read_intr;
		return;
	}
	end_request(1);
	do_hd_request();
}

static void write_intr(void)
{
	if (win_result()) {
		bad_rw_intr();
		do_hd_request();
		return;
	}
	if (--CURRENT->nr_sectors) {
		CURRENT->sector++;
		CURRENT->buffer += 512;
		do_hd = &write_intr;
		port_write(HD_DATA,CURRENT->buffer,256);
		return;
	}
	end_request(1);
	do_hd_request();
}

static void recal_intr(void)
{
	if (win_result())
		bad_rw_intr();
	do_hd_request();
}

void do_hd_request(void)
{
	int i,r;
	unsigned int block,dev;
	unsigned int sec,head,cyl;
	unsigned int nsect;

	INIT_REQUEST;
	dev = MINOR(CURRENT->dev);
	block = CURRENT->sector;
	if (dev >= 5*NR_HD || block+2 > hd[dev].nr_sects) {
		end_request(0);
		goto repeat;
	}
	block += hd[dev].start_sect;
	dev /= 5;
	__asm__("divl %4":"=a" (block),"=d" (sec):"0" (block),"1" (0),
		"r" (hd_info[dev].sect));
	__asm__("divl %4":"=a" (cyl),"=d" (head):"0" (block),"1" (0),
		"r" (hd_info[dev].head));
	sec++;
	nsect = CURRENT->nr_sectors;
	if (reset) {
		reset = 0;
		recalibrate = 1;
		reset_hd(CURRENT_DEV);
		return;
	}
	if (recalibrate) {
		recalibrate = 0;
		hd_out(dev,hd_info[CURRENT_DEV].sect,0,0,0,
			WIN_RESTORE,&recal_intr);
		return;
	}	
	if (CURRENT->cmd == WRITE) {
		hd_out(dev,nsect,sec,head,cyl,WIN_WRITE,&write_intr);
		for(i=0 ; i<3000 && !(r=inb_p(HD_STATUS)&DRQ_STAT) ; i++)
			/* nothing */ ;
		if (!r) {
			bad_rw_intr();
			goto repeat;
		}
		port_write(HD_DATA,CURRENT->buffer,256);
	} else if (CURRENT->cmd == READ) {
		hd_out(dev,nsect,sec,head,cyl,WIN_READ,&read_intr);
	} else
		panic("unknown hd-command");
}

void hd_init(void)
{
	blk_dev[MAJOR_NR].request_fn = DEVICE_REQUEST;
	set_intr_gate(0x2E,&hd_interrupt);
	outb_p(inb_p(0x21)&0xfb,0x21);
	outb(inb_p(0xA1)&0xbf,0xA1);
}
